python/lsst/daf/butler/core/datastoreRecordData.py

```diff
@@ -28,32 +28,68 @@
 import dataclasses
 import uuid
 from collections import defaultdict
-from typing import TYPE_CHECKING, AbstractSet, Dict, List, Optional
+from typing import TYPE_CHECKING, AbstractSet, Any, Dict, List, Optional, Set
 
-from pydantic import BaseModel
+from lsst.utils import doImportType
+from lsst.utils.introspection import get_full_type_name
+from pydantic import BaseModel, validator
 
 from .datasets import DatasetId
 from .dimensions import DimensionUniverse
-from .storedFileInfo import SerializedStoredDatastoreItemInfo, StoredDatastoreItemInfo
+from .storedFileInfo import StoredDatastoreItemInfo
 
 if TYPE_CHECKING:
 
     from ..registry import Registry
-    from ..registry.interfaces import DatasetIdRef
+
+_Record = Dict[str, Any]
+
+
+def _str_to_uuid(dataset_id: Any) -> uuid.UUID:
+    """Convert dataset_id from string back to UUID, e.g. after JSON
+    round-trip.
+    """
+    if isinstance(dataset_id, str):
+        return uuid.UUID(dataset_id)
+    else:
+        assert isinstance(dataset_id, uuid.UUID), "Expecting UUIDs"
+        return dataset_id
```

Comment:
The issue here is JSON which stores UUIDs as strings and reads string back, and regular construction from datastore should always use true UUIDs. The problem is that both of those cases use regular constructor. We also have a `direct()` method that we use when restoring quantum graph from saved data. If we could say that JSON deserialization have to always use `direct()` instead of `__init__` than we can skip the conversion in `__init__` but always do it in `direct()` (but then we loose validation when reading from JSON).

Reply:
I removed validators, constructor now accepts UUIDs only, and `direct()` method converts strings to UUIDs, similarly to how it's done for `DatasetRef` class.